Rust: Basic control flow graph setup

paldepind opened the pull request:
This is an attempt to add the basic setup/boilerplate needed to construct a CFG.

The PR is based on looking at how the CFG library is used for Swift, [Kaleidoscope](https://github.com/aibaars/codeql-kaleidoscope), Ruby, and C#. I've mixed and matched a bit form these, but mostly taken stuff from Swift (because it also uses codegen) and Kaleidoscope (because it's the simplest setup). I've erred on the side of simplicity to avoid adding stuff that we don't need (right away at least).

hvitved commented on rust/ql/lib/codeql/rust/controlflow/BasicBlocks.qll:
```diff
@@ -0,0 +1,262 @@
+ * A basic block, that is, a maximal straight-line sequence of control flow nodes
+ * without branches or joins.
+ */
+class BasicBlock extends TBasicBlockStart {
```
final

paldepind replied:
`BasicBlock` is extended further down in the file so I did
```ql
final class BasicBlock = BasicBlockImpl;

private class BasicBlockImpl extends TBasicBlockStart { ... }
```
similar to what you suggested for `SuccessorType`.

hvitved commented on rust/ql/lib/codeql/rust/controlflow/internal/Scope.qll:
```diff
@@ -0,0 +1,36 @@
+
+  /** Gets the nearest enclosing parent of `ast` that is an `AstNode`. */
+  cached
+  AstNode getParentOfAst(AstNode ast) { result = getParentOfAstStep*(getImmediateParent(ast)) }
```
This predicate is only used in this file, so I would make it `private` and non-`cached`.

paldepind replied:
And can I then also remove the surrounding `private module`?

hvitved replied:
Yes.

hvitved commented on rust/ql/lib/codeql/rust/controlflow/ControlFlowGraph.qll:
```diff
@@ -0,0 +1,36 @@
+ *
+ * Only nodes that can be reached from an entry point are included in the CFG.
+ */
+class CfgNode extends Node {
```
`final`. This also means you can move all `final` annotations on predicates.

hvitved commented on rust/ql/lib/codeql/rust/controlflow/ControlFlowGraph.qll:
```diff
@@ -0,0 +1,36 @@
+private import internal.Completion
+private import internal.SuccessorType
+private import codeql.rust.controlflow.BasicBlocks
+import internal.Scope
```
This exposes the `abstract` class `CfgScope`, which users could then in principle extend. So I think it would be better to do
```ql
private import internal.Scope as Scope

final class CfgScope = Scope::CfgScope;
```

paldepind replied:
I've done as suggested. But I wonder if it would make sense to do it inside `Scope.qll` instead?

hvitved commented on rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImplSpecific.qll:
```diff
@@ -0,0 +1,63 @@
+private import SuccessorType as ST
+private import Scope as Scope
+
+module CfgInput implements InputSig<Rust::Location> {
```
I would just move this into `ControlFlowGraphImpl.qll` and get rid of `ControlFlowGraphImplSpecific.qll`.

paldepind replied:
I picked this separation up from `swift`, but yeah I agree that it doesn't seem to meaningful.

hvitved commented on rust/ql/lib/codeql/rust/controlflow/internal/PrintCFG.ql:
```diff
@@ -0,0 +1,51 @@
+/**
```
Nit: Should be renamed to `PrintCfg.ql` (we only upper-case the first letter in acronyms in file/class/predicate names; so e.g. `Ssa` and `Cfg` instead of `SSA` and `CFG`).

hvitved commented on rust/ql/lib/codeql/rust/controlflow/internal/PrintCFG.ql:
```diff
@@ -0,0 +1,51 @@
+
+private predicate selectedSourceColumnAlias = selectedSourceColumn/0;
+
+module ViewCfgQueryInput implements Impl::ViewCfgQueryInputSig<File> {
```
private

hvitved replied:
(not that it really matters in a `.ql` file).

hvitved commented on rust/ql/lib/codeql/rust/controlflow/internal/SuccessorType.qll:
```diff
@@ -0,0 +1,46 @@
+  TReturnSuccessor()
+
+/** The type of a control flow successor. */
+class SuccessorType extends TSuccessorType {
```
Since this class is exposed, I would do
```ql
private abstract class SuccessorTypeImpl extends TSuccessorType {
  /** Gets a textual representation of successor type. */
  abstract string toString();
}

final class SuccessorType = SuccessorTypeImpl;

// ...

private abstract class ConditionalSuccessorImpl extends SuccessorType {
 // ...
}

final class ConditionalSuccessor = ConditionalSucessorImpl;

// etc (make all public classes final)
```

hvitved commented on rust/ql/test/library-tests/controlflow/CFG.ql:
```diff
@@ -0,0 +1,13 @@
+/**
+ * @kind graph
```
Rename to `Cfg.ql`. Also, I suggest removing `@kind graph` as, in my experience, it actually hurts more than it helps; for example, nodes with multiple edges between them get collapsed in this format

hvitved commented on rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:
```diff
@@ -0,0 +1,47 @@
+    // An edge from the then branch to the last node
+    last(super.getThen(), pred, c) and
+    succ = this and
+    completionIsSimple(c)
```
Should rather be `completionIsNormal(c)`.

hvitved commented on rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:
```diff
@@ -0,0 +1,47 @@
+    // An edge from the else branch to the last node
+    last(super.getElse(), pred, c) and
+    succ = this and
+    completionIsSimple(c)
```
Same

hvitved commented on rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:
```diff
@@ -0,0 +1,47 @@
+class IfTree extends PostOrderTree instanceof If {
+  override predicate first(AstNode node) { first(super.getCondition(), node) }
+
+  override predicate propagatesAbnormal(AstNode child) { none() }
```
I think this should be `child = [super.getCondition(), super.getThen(), super.getElse()]`; for example `return`s are also considered "abnormal".

hvitved reviewed (changes requested):
Looks great! A few (mostly minor) comments.

paldepind commented:
I've addressed the comments 👍 

I think we have the basic foundation for the CFG now. Looking at [the result of the test](https://github.com/github/codeql/pull/17415/files#diff-7c2877201f76e8c40696b62f56dd2dbe6e6f97241531952c499f341a069fa321) we are at the very least creating some nodes and edges :) My guess it that they will still require some tweaking to be entirely right though.
